src/libfsm/print/vmasm.c

```diff
@@ -0,0 +1,342 @@
+			}
+
+			curr_st = op->ir_state;
+		}
+
+		switch (op->instr) {
+		case VM_OP_STOP:
+			{
+				const char *jmp_op;
+
+				if (op->u.stop.end_bits == VM_END_SUCC) {
+					unsigned end_st = op->ir_state - ir->states;
+					switch (dialect) {
+					case AMD64_ATT:
+						fprintf(f, "\tmovl    $0x%02x, %%%s\n", end_st, ret_reg);
+						break;
+					case AMD64_NASM:
+						fprintf(f, "\tMOV   %s, %02xh\n", ret_reg, end_st);
+						break;
+					}
+				} else {
+					switch (dialect) {
+					case AMD64_ATT:
+						fprintf(f, "\tmovl    $-1, %%%s\n", ret_reg);
+						break;
+					case AMD64_NASM:
+						fprintf(f, "\tMOV   %s, -1\n", ret_reg);
+						break;
+					}
+				}
+
+				if (op->cmp == VM_CMP_ALWAYS && op->next == NULL) {
+					fprintf(f, "\t%c elided jmp to .finish\n", comment);
+				} else {
+					if (op->cmp != VM_CMP_ALWAYS) {
+						switch (dialect) {
+						case AMD64_ATT:
+							fprintf(f, "\tcmpl    $0x%02x,%%%s\n", (unsigned)op->cmp_arg, chr_reg);
+							break;
+						case AMD64_NASM:
+							fprintf(f, "\tCMP   %s,%02xh\n", chr_reg, (unsigned)op->cmp_arg);
+							break;
+						}
+					}
+
+					switch (op->cmp) {
+					case VM_CMP_ALWAYS: jmp_op = (dialect == AMD64_ATT) ? "jmp    " : "JMP  "; break;
+					case VM_CMP_LT:     jmp_op = (dialect == AMD64_ATT) ? "jb     " : "JB   "; break;
+					case VM_CMP_LE:     jmp_op = (dialect == AMD64_ATT) ? "jbe    " : "JBE  "; break;
+					case VM_CMP_GE:     jmp_op = (dialect == AMD64_ATT) ? "jae    " : "JAE  "; break;
+					case VM_CMP_GT:     jmp_op = (dialect == AMD64_ATT) ? "ja     " : "JA   "; break;
+					case VM_CMP_EQ:     jmp_op = (dialect == AMD64_ATT) ? "je     " : "JE   "; break;
+					case VM_CMP_NE:     jmp_op = (dialect == AMD64_ATT) ? "jne    " : "JNE  "; break;
+					}
+
+					fprintf(f, "\t%-3s .finish\n", jmp_op);
+				}
+			}
+			break;
+
+		case VM_OP_FETCH:
+			{
+				if (op->u.fetch.end_bits == VM_END_SUCC) {
+					unsigned end_st = op->ir_state - ir->states;
+
+					switch (dialect) {
+					case AMD64_ATT:
+						fprintf(f, "\tmovl    $0x%02x,%%%s\n", (unsigned)end_st, ret_reg);
+						break;
+
+					case AMD64_NASM:
+						fprintf(f, "\tMOV   %s, %02xh\n", ret_reg, end_st);
+						break;
+					}
+				} else {
+
+					switch (dialect) {
+					case AMD64_ATT:
+						fprintf(f, "\tmovl    $-1, %%%s\n", ret_reg);
+						break;
+
+					case AMD64_NASM:
+						fprintf(f, "\tMOV   %s, -1\n", ret_reg);
+						break;
+					}
+				}
+
+
+				switch (dialect) {
+				case AMD64_ATT:
+					fprintf(f, "\tcmpq    %%%s,%%%s\n", stp_reg, stn_reg);
+					fprintf(f, "\tje      .finish\n");
+					fprintf(f, "\tmovzbl  (%%%s), %%%s\n", stp_reg, chr_reg);
+					fprintf(f, "\taddq    $1, %%%s\n", stp_reg);
+					break;
+
+				case AMD64_NASM:
+					fprintf(f, "\tCMP   %s, %s\n", stp_reg, stn_reg);
+					fprintf(f, "\tJE    .finish\n");
+					fprintf(f, "\tMOVZX %s, BYTE [%s]\n", chr_reg, stp_reg);
+					fprintf(f, "\tADD   %s, 1\n", stp_reg);
+					break;
+				}
```

Comment:
With so few instruction types, maybe it'd suit us to have like `emit_*()` for each instruction, passed whatever arguments suit it, and switch on the dialect in there. I did something similar with factoring together the various assembly dialects for tendra, and that worked out pretty well. It might suit us for JIT, too - except passing `FILE *` is a bit awkward there.

Reply:
I was thinking this, but decided that I'd rather keep it simple and just get it done.